### PR TITLE
[JAX] Adapt latest JAX/PAX image

### PR DESCRIPTION
**Pull request by zlsh80826**
1. custom primitives requires the same shape for fwd input and bwd corresponding gradients. We used shape = [1] for dummy inputs in the past. This PR changed the input shape to be match as the bwd return output.
2. Latest JAX changed their layernorm fusion kernel, so the original ref code computed the `scale * y + bias` in bfloat16 which leads higher difference with TE's layernorm after 03-14 image. This PR makes the reference code computes `scale * y + bias` in higher precision and then cast back to the bfloat16, which aligns [TE's implementations](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/layer_norm/ln_fwd_kernels.cuh#L120-L125) (fp32 register computation and cast to F16 before saving to global memory).
3. WAR `jax.distributed.init` issue by not providing `local_device_ids`, which forces JAX to query devices and call `cuinit`.

### CODE REVIEW

**Comment by zlsh80826** [comment]
PIpeline: 13991833

File: examples/jax/encoder/test_multiprocessing_encoder.py [hunk 1]
```diff
@@ -315,8 +315,7 @@ def train_and_evaluate(args):
 
     jax.distributed.initialize(coordinator_address=args.coordinator_address,
                                num_processes=args.num_process,
-                               process_id=args.process_id,
-                               local_device_ids=args.process_id)
+                               process_id=args.process_id)
```
Comment (nouiz) [inline examples/jax/encoder/test_multiprocessing_encoder.py:318]: Can you explain more why this changes is needed?

**Review by nouiz — APPROVED** [review]
LGTM. Just had a question.

File: examples/jax/encoder/test_multiprocessing_encoder.py (hunk 1 above)
Comment (zlsh80826) [reply examples/jax/encoder/test_multiprocessing_encoder.py:318]: It is a WAR. There are bugs that JAX doesn't call `cuInit` before querying the device compute capability and causing the `cuda_not_initialized` error. However, we can WAR it by not providing the local_devices_ids, so that JAX will first query the number of devices and implicit calling `cuInit` and make the following device compute capability call work.

**Comment by nouiz** [comment]
Note, for the JAX release this week, we need this PR.
Make sure the CI run and it can be merged.

File: examples/jax/encoder/test_multiprocessing_encoder.py (hunk 1 above)
Comment (nouiz) [reply examples/jax/encoder/test_multiprocessing_encoder.py:318]: Thanks. An upstream fix got merged, but one JAX releases was done. So to be compatible with all JAX version, lets merge this.

**Comment by denera** [comment]
@ptrendx The CI I triggered earlier didn't get reported to GitHub for some reason but it's been running (pipeline #13991833) and I can see that distributed unit tests have all passed. Should we merge or do you wanna wait for yours to finish?

**Comment by ptrendx** [comment]
The pipeline you are referring to is the one Reese triggered. There was some hiccup in Jenkins which resulted in pipeline not being created. I see now that there were no further changes after Reese triggered the CI so I'm fine with just that one, although I see that the A100 test did not run there yet - I think we should wait for it to finish, right?

**Comment by denera** [comment]
Ah I see now, mine didn't go through for some reason and I didn't get a notice about it. 🤷🏻‍♂️

I believe the only changes in this PR are for distributed runs so not waiting for the A100 single GPU tests would be fine.

**Comment by zlsh80826** [comment]
I found the WAR for multi-processes didn't work on all environments, so I reverted it. Since the latest jax image fixes the cuInit issue, so it is safe to revert it. The passed pipeline is 14042250. Thanks @denera 's quick merge!

**Comment by nouiz** [comment]
> I found the WAR for multi-processes didn't work on all environments, so I reverted it. Since the latest jax image fixes the cuInit issue, so it is safe to revert it. The passed pipeline is 14042250. Thanks @denera 's quick merge!

Did it cause regression? Or just that it doesn't fix all cases?
It isn't safe to just remove the WAR as the latest JAX releases isn't fixed. So right now, people can't use TE with the latest JAX releases. That isn't great.

**Comment by zlsh80826** [comment]
> > I found the WAR for multi-processes didn't work on all environments, so I reverted it. Since the latest jax image fixes the cuInit issue, so it is safe to revert it. The passed pipeline is 14042250. Thanks @denera 's quick merge!
> 
> Did it cause regression? Or just that it doesn't fix all cases? It isn't safe to just remove the WAR as the latest JAX releases isn't fixed. So right now, people can't use TE with the latest JAX releases. That isn't great.

It didn't fix all cases.
With the WAR: all images fails with num_gpus >= 2
Without the WAR: images which newer than 04-05 work
